enhancements/cgu-precaching-additional-images.md

```diff
@@ -0,0 +1,308 @@
+(SNO+workers). This EP will also target SNOs only and the capability for multi-node configurations
+will be scoped as a future enhancement.
+
+The following items are out of scope for this EP:
+- There will be no image registry or UI for inspecting the cached images.
+- Management of the user application is not included. This epic comprises only pre-caching
+user-specified images in CRIO.
+
+
+## Proposal
+
+The proposed solution is to leverage the existing pre-caching job on TALM to pre-cache the additional
+user-specified images. This job is currently used to pre-cache the OpenShift platform image and operators
+on the managed clusters. The job runs on the clusters and pulls the requested set of missing images/layers.
+
+This proposal further allows the application of the existing Cluster Group Upgrade (GGU) custom resource (CR)
+to enable the user to pre-cache the additional images.
+
+The following changes are suggested:
+
+1. Add a new `AdditionalImages` field to the `PrecachingSpec` struct in
+`cluster-group-upgrades-operator/api/v1alpha1/clustergroupupgrade_types.go` as shown below.
+    ```golang
+    ...
+    // PrecachingSpec defines the pre-caching software spec derived from policies
+    type PrecachingSpec struct {
+        PlatformImage                string   `json:"platformImage,omitempty"`
+        OperatorsIndexes             []string `json:"operatorsIndexes,omitempty"`
+        OperatorsPackagesAndChannels []string `json:"operatorsPackagesAndChannels,omitempty"`
+        ExcludePrecachePatterns      []string `json:"excludePrecachePatterns,omitempty"`
+        AdditionalImages    []string `json:"additionalImages,omitempty"`
+    }
+   ...
+
+2. Apply the operator-sdk tool to update the TALM custom resource definitions, manifests and auto-generated
+golang source files.
+
+3. Extend the `controllers/managedClusterResources.go` and
+`cluster-group-upgrades-operator/controllers/precache.go` source files to handle reconciliation
+control logic pertaining to the `additionalImages` object.
+
+4. Extend the pre-cache template source file `controllers/templates/precache-templates.go` with the
+`additionalImages` field.
+
+5. Refactor the `cluster-group-upgrades-operator/pre-cache/precache.sh` and
+`cluster-group-upgrades-operator/pre-cache/pull` scripts to check for additional user images and pull
+them accordingly. The existing image pulling functionality should be refactored in order to promote
+code reusability and avoid code duplication.
+
+### Workflow Description
+
+The general workflow for pre-caching user specified images is as follows:
+`ClusterSelected -> Validated -> PrecacheSpecValid -> PrecachingSucceeded -> <TALM workflow continues>`.
+
+Detailed user and worklow steps are outlined below.
+
+1. User creates a CGU CR with the `preCaching` field set to `true` and prepares a ConfigMap (named
+`cluster-group-upgrade-overrides`) with the list of additional images to be pre-cached as shown in
```

Comment:
Rather than have an implicit relationship between the configmap named `cluster-group-upgrade-overrides` can we add a field to the spec where the user explicitly links to the precache configmap? I think this has a couple of benefits
1. The user can see the explicit inclusion of the additional pre-cache objects. When it is implicit the user needs to know the name of the configmap in order to find it (debug scenarios)
1. Ensures the additional pre-cache objects are used only when intended. For example an old/stale configmap in the namespace from a prior run. Or multiple CGU in the same namespace for different functions, sets of clusters, etc.

Reply:
We can indeed add a new field to the CGU spec,  e.g. `preCachingAdditionalImagesRef`, which would reference the ConfigMap containing the additional images to be pre-cached. However, this would create an anti-pattern of extending the pre-caching specification fields in the existing CGU by repeatedly entering the prefix “preCaching”. An alternative and future-proof design option would entail creating a new CR for the pre-caching specification and reference this in the existing CGU. The new pre-caching CR would serve to implement the existing `PrecachingSpec` as an object and we would one day deprecate the ConfigMap override behaviour. 

For example:
```yaml
kind: ClusterGroupUpgrade
metadata:
  name: cgu
  namespace: default
spec:
  clusters:
  - sno1
  - sno3
  - sno7
  enable: false
  preCaching: true
  preCachingSpecRef:
    name: foobar
    namespace: foobar-ns
  backup: false
  managedPolicies:
  - policy0-common-config-policy
  remediationStrategy:
    maxConcurrency: 4
    timeout: 240
```
New CR:
```yaml
apiVersion: ran.openshift.io/v1alpha1
kind: Precaching
metadata:
  name: foobar
  namespace: foobar-ns
spec:
  platformImage: quay.io/openshift-release-dev/ocp-release@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
  operatorsIndexes:
    - registry.example.com:5000/custom-redhat-operators:1.0.0
  operatorsPackagesAndChannels:
    - local-storage-operator: stable
    - performance-addon-operator: 4.9
    - ptp-operator: stable
    - sriov-network-operator: stable
  excludePrecachePatterns:
    - aws
    - alibaba
    - azure
    - vsphere
  additionalImages:
    - quay.io/foobar/application1@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
    - quay.io/foobar/application2@sha256:3d5800123dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47adfaef
    - quay.io/foobar/applicationN@sha256:4fe1334adfafadsf987123adfffdaf1243340adfafdedga0991234afdadfsa09
```

Note that the fields are optional and can be omitted. TALM will infer the necessary pre-caching images.

Reply:
@jc-rh what are your thoughts on this?

Reply:
+1 on abandoning the configmap in favor of a pre-caching CR. Also will need to move "SpaceRequired" here.

Reply:
There could be cases where we want to exclude images (sha or by tag) instead of just "grepping" by a pattern.  Or maybe the grep excludes an image that is needed.

Make sense to create a new field called excludeImages? and add the images to be excluded explicitly instead of relying on a regular expression?

Below the current way to exclude images:

```
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-group-upgrade-overrides
  namespace: ztp-group-du-sno
data:
#  operators.indexes: |
#    jumphost.inbound.vz.bos2.lab:5000/ocp4/redhat/redhat-operator-index:v4.10
  # Review these filters/patterns before applying it!
  excludePrecachePatterns: |
    quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:df2ca7f54a8d34f568a92c077c56ed8983fffc51146b55964971c1f970288aaa
    quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:7c330b88fb2d0bf06b02892745c7fc3096b6325a49c854f3a7d80a87048f6401
    quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:a4934b12e3698a384826bae660ebfb84a30f68e0519fee949ebeb4f4074f26c6
    quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:ea1f013e3a6fca4d1d7777c455f0e34be49bebd2cb1ccac16b140d49196916e3
...COMENTED...
 quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:938310aeedba130b66204b062b7d86d562bdff2d6767fd643176cea3754e0d37
```

Reply:
I also like the idea of having a separated pre-caching CR, it would provide further control/flexibility over this feature which is currently needed.

Reply:
One of the benefits of TALM's current pre-caching is that TALM reviews the set of Policies that are attached and derives the set of images to be precached from those. With the full PreCaching CR above the responsibility has been shifted to the user to identify OCP version, operator index, etc. Perhaps there is a middle-ground here between the two where the CGU has a precaching struct in the spec and the image ref would live there:
```
kind: ClusterGroupUpgrade
metadata:
  name: cgu
  namespace: default
spec:
  precache:
    enable: true
    additionalImagesRef: <cm name>
    excludePattern: ....
```
The current `spec.preCaching: true` would be deprecated in favor of the `spec.precache.enable`.

Reply:
The proposed pre-caching CR would not change the current functionality of TALM, i.e.  TALM will still  infer the necessary pre-caching images. Note that the fields in the new pre-caching CR are optional and serve as an override. The objective is to one-day deprecate the current TALM behaviour of using a ConfigMap to override the pre-caching specifications.

Reply:
> @jc-rh what are your thoughts on this?

Why are you asking me? I like it of course as it was my idea :smirk:

Reply:
Ah, so the `platformImage`, `operatorsIndexes`, and `operatorsPackagesAndChannels` fields are overrides of the TALM derived values...
I like the idea of precaching having an explicit structure for defining its content. That should help make clear to the user what the function is and provide preliminary validation against the CRD.
That said, the behavior/meaning of the first three fields is fundamentally different than the last two. The first three are automatically derived unless overrides are specified here. Perhaps these can be re-classified as "override..." to make that clear?

Reply:
Sounds good @imiller0. I like the suggestion of re-classifying `platformImage`, `operatorsIndexes`, and `operatorsPackagesAndChannels` fields as override. Something like this:

```yaml
apiVersion: ran.openshift.io/v1alpha1
kind: PrecacheSpecRef
metadata:
  name: foobar
spec:
  overrides:
    platformImage: quay.io/openshift-release-dev/ocp-release@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
    operatorsIndexes:
      - registry.example.com:5000/custom-redhat-operators:1.0.0
    operatorsPackagesAndChannels:
      - local-storage-operator: stable
      - performance-addon-operator: 4.9
      - ptp-operator: stable
      - sriov-network-operator: stable
  excludePrecachePatterns:
    - aws
    - alibaba
    - azure
    - vsphere
  additionalImages:
    - quay.io/foobar/application1@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
    - quay.io/foobar/application2@sha256:3d5800123dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47adfaef
    - quay.io/foobar/applicationN@sha256:4fe1334adfafadsf987123adfffdaf1243340adfafdedga0991234afdadfsa09
```

Reply:
I like that better. It helps clarify the roles of the fields and that the 3 overrides are "optional" for most users.

One note. The kind should be something like PrecacheSpec and then the reference to it from the CGU would be the precacheSpecRef.

Reply:
To me, the overriding behavior is more of a TALM implementation choice, rather than an API thing. In the future, if TALM does start populate the other two fields, we would need to decide the override behavior for them. That decision probably shouldn't be tied to the API definition of this object. I would rather keep it flat and reflect the override aspect with the field name in the CGU spec, e.g. precacheSpecOverrideRef. Also, this looks a bit awkward in the CGU status which needs to show the actual precaching spec we use in the end.

Reply:
Sharat convinced me it could be misleading to customers calling the whole thing an override. In that case, we will probably need a different name as there is already a precaching/spec in CGU status with all the same fields but without overrides. The best I can come up with is PrecachingConfig.

Reply:
Yep, I think we should get rid of the override. It was for development / debug in the first place. 
IMHO, PrecachingConfig nails it

Reply:
Thank you all for your valuable feedback. I have taken the comments and suggestions into account.